File: experiments/runs/2026-09-14_20-55-51_claim-office-example-mapping_baseline-inline-tdd-v1.1-local-git-cc_opus-5-no-thinking-2/src/policy.ts
import { isKnownType } from './catalog.js';
import type { Item } from './premium.js';
import { insuranceSum, policyCap, settleDamage } from './claim.js';

export interface Damage {
  itemType: string;
  amount: number;
}

export interface Incident {
  cause: string;
  damages: Damage[];
}

export interface ClaimResult {
  payout: number;
  remainingCap: number;
}

export class ClaimRejectedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ClaimRejectedError';
  }
}

/**
 * A policy created by a quote step. It owns the insured items and the
 * remaining cap, which successive claims draw down.
 */
export class Policy {
  readonly insuranceSum: number;
  private remaining: number;

  constructor(private readonly items: Item[]) {
    this.insuranceSum = insuranceSum(items);
    this.remaining = policyCap(items);
  }

  get remainingCap(): number {
    return this.remaining;
  }

  /**
   * Settles an incident against this policy. The whole claim is rejected if
   * any damage entry cannot be matched to a distinct insured item, or carries
   * a negative amount, so a rejected claim leaves the cap untouched.
   */
  claim(incident: Incident): ClaimResult {
    const available = [...this.items];
    let desired = 0;

    for (const damage of incident.damages) {
      if (damage.amount < 0) {
        throw new ClaimRejectedError(`negative damage amount: ${damage.amount}`);
      }
      if (!isKnownType(damage.itemType)) {
        throw new ClaimRejectedError(`unknown item type: ${damage.itemType}`);
      }
      const index = available.findIndex((item) => item.type === damage.itemType);
      if (index === -1) {
        throw new ClaimRejectedError(`item not covered by this policy: ${damage.itemType}`);
      }
      const [insured] = available.splice(index, 1);
      desired += settleDamage(insured, damage.amount);
    }

    const payout = Math.min(desired, this.remaining);
    this.remaining -= payout;
    return { payout, remainingCap: this.remaining };
  }
}

File: experiments/runs/2026-09-20_07-22-29_claim-office-java-example-mapping_baseline-inline-tdd-v1-cc_opus-5-no-thinking/src/main/java/Policy.java
import java.math.BigDecimal;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** A policy created by a quote, settling claims against its remaining cap. */
final class Policy {

    private final List<Item> items;
    private final int insuranceSum;
    private BigDecimal remainingCap;

    Policy(List<Item> items) {
        this.items = List.copyOf(items);
        int sum = 0;
        for (Item item : this.items) {
            sum += item.insuranceValue();
        }
        this.insuranceSum = sum;
        this.remainingCap = BigDecimal.valueOf(2L * sum);
    }

    int insuranceSum() {
        return insuranceSum;
    }

    int remainingCap() {
        return Claims.roundPayout(remainingCap);
    }

    /** Settles one incident against this policy and returns the payout in whole G. */
    int settle(List<Damage> damages) {
        List<Item> damagedItems = matchDamagedItems(damages);
        BigDecimal desired = BigDecimal.ZERO;
        for (int i = 0; i < damages.size(); i++) {
            BigDecimal amount = BigDecimal.valueOf(damages.get(i).amount());
            desired = desired.add(Claims.itemPayout(damagedItems.get(i), amount));
        }
        BigDecimal granted = desired.min(remainingCap);
        remainingCap = remainingCap.subtract(granted);
        return Claims.roundPayout(granted);
    }

    private List<Item> matchDamagedItems(List<Damage> damages) {
        Map<String, Deque<Item>> available = new LinkedHashMap<>();
        for (Item item : items) {
            available.computeIfAbsent(item.type(), type -> new ArrayDeque<>()).add(item);
        }
        List<Item> matched = new ArrayList<>();
        for (Damage damage : damages) {
            if (damage.amount() < 0) {
                throw new ClaimOfficeException("negative damage amount: " + damage.amount());
            }
            Deque<Item> candidates = available.get(damage.itemType());
            if (candidates == null || candidates.isEmpty()) {
                throw new ClaimOfficeException("item not covered by the policy: " + damage.itemType());
            }
            matched.add(candidates.remove());
        }
        return matched;
    }
}
